In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import random
from wordcloud import WordCloud

###Loading of the Zomato dataset.

##Preprocessing of the data

###Exploring the data

In [4]:
pd.read_csv("/zomato_restaurants_in_India.csv")

res_id                            name      establishment  \
0        3400299                     Bikanervala    ['Quick Bites']   
1        3400005  Mama Chicken Mama Franky House    ['Quick Bites']   
2        3401013                   Bhagat Halwai    ['Quick Bites']   
3        3400290                   Bhagat Halwai    ['Quick Bites']   
4        3401744     The Salt Cafe Kitchen & Bar  ['Casual Dining']   
...          ...                             ...                ...   
211939   3202251  Kali Mirch Cafe And Restaurant  ['Casual Dining']   
211940   3200996                      Raju Omlet    ['Quick Bites']   
211941  18984164                The Grand Thakar  ['Casual Dining']   
211942   3201138                          Subway    ['Quick Bites']   
211943  18879846     Freshco's - The Health Cafe           ['Café']   

                                                      url  \
0       https://www.zomato.com/agra/bikanervala-khanda...   
1       https://www.zomato.com/agra/mama-chicken-mama-...   
2       https://www.zomato.com/agra/bhagat-halwai-2-sh...   
3       https://www.zomato.com/agra/bhagat-halwai-civi...   
4       https://www.zomato.com/agra/the-salt-cafe-kitc...   
...                                                   ...   
211939  https://www.zomato.com/vadodara/kali-mirch-caf...   
211940  https://www.zomato.com/vadodara/raju-omlet-kar...   
211941  https://www.zomato.com/vadodara/the-grand-thak...   
211942  https://www.zomato.com/vadodara/subway-1-akota...   
211943  https://www.zomato.com/vadodara/freshcos-the-h...   

                                                  address      city  city_id  \
0       Kalyani Point, Near Tulsi Cinema, Bypass Road,...      Agra       34   
1             Main Market, Sadar Bazaar, Agra Cantt, Agra      Agra       34   
2       62/1, Near Easy Day, West Shivaji Nagar, Goalp...      Agra       34   
3       Near Anjana Cinema, Nehru Nagar, Civil Lines, ...      Agra       34   
4             1C,3rd Floor, Fatehabad Road, Tajganj, Agra      Agra       34   
...                                                   ...       ...      ...   
211939  Manu Smriti Complex, Near Navrachna School, GI...  Vadodara       32   
211940  Mahalaxmi Apartment, Opposite B O B, Karoli Ba...  Vadodara       32   
211941  3rd Floor, Shreem Shalini Mall, Opposite Conqu...  Vadodara       32   
211942  G-2, Vedant Platina, Near Cosmos, Akota, Vadodara  Vadodara       32   
211943  Shop 7, Ground Floor, Opposite Natubhai Circle...  Vadodara       32   

           locality   latitude  longitude  ... price_range  currency  \
0          Khandari  27.211450  78.002381  ...           2       Rs.   
1        Agra Cantt  27.160569  78.011583  ...           2       Rs.   
2          Shahganj  27.182938  77.979684  ...           1       Rs.   
3       Civil Lines  27.205668  78.004799  ...           1       Rs.   
4           Tajganj  27.157709  78.052421  ...           3       Rs.   
...             ...        ...        ...  ...         ...       ...   
211939    Fatehgunj  22.336931  73.192356  ...           2       Rs.   
211940   Karelibaug  22.322455  73.197203  ...           1       Rs.   
211941     Alkapuri  22.310563  73.171163  ...           2       Rs.   
211942        Akota  22.270027  73.143068  ...           2       Rs.   
211943     Vadiwadi  22.309935  73.158768  ...           2       Rs.   

                                               highlights aggregate_rating  \
0       ['Lunch', 'Takeaway Available', 'Credit Card',...              4.4   
1       ['Delivery', 'No Alcohol Available', 'Dinner',...              4.4   
2       ['No Alcohol Available', 'Dinner', 'Takeaway A...              4.2   
3       ['Takeaway Available', 'Credit Card', 'Lunch',...              4.3   
4       ['Lunch', 'Serves Alcohol', 'Cash', 'Credit Ca...              4.9   
...                                                   ...              ...   
211939  ['Dinner', 'Cash', 'Lunch', 'Delivery', 'Ind

In [6]:
data  = pd.read_csv("/zomato_restaurants_in_India.csv")

In [7]:
data.shape

(211944, 26)

In [8]:
data.tail()

res_id                            name      establishment  \
211939   3202251  Kali Mirch Cafe And Restaurant  ['Casual Dining']   
211940   3200996                      Raju Omlet    ['Quick Bites']   
211941  18984164                The Grand Thakar  ['Casual Dining']   
211942   3201138                          Subway    ['Quick Bites']   
211943  18879846     Freshco's - The Health Cafe           ['Café']   

                                                      url  \
211939  https://www.zomato.com/vadodara/kali-mirch-caf...   
211940  https://www.zomato.com/vadodara/raju-omlet-kar...   
211941  https://www.zomato.com/vadodara/the-grand-thak...   
211942  https://www.zomato.com/vadodara/subway-1-akota...   
211943  https://www.zomato.com/vadodara/freshcos-the-h...   

                                                  address      city  city_id  \
211939  Manu Smriti Complex, Near Navrachna School, GI...  Vadodara       32   
211940  Mahalaxmi Apartment, Opposite B O B, Karoli Ba...  Vadodara       32   
211941  3rd Floor, Shreem Shalini Mall, Opposite Conqu...  Vadodara       32   
211942  G-2, Vedant Platina, Near Cosmos, Akota, Vadodara  Vadodara       32   
211943  Shop 7, Ground Floor, Opposite Natubhai Circle...  Vadodara       32   

          locality   latitude  longitude  ... price_range  currency  \
211939   Fatehgunj  22.336931  73.192356  ...           2       Rs.   
211940  Karelibaug  22.322455  73.197203  ...           1       Rs.   
211941    Alkapuri  22.310563  73.171163  ...           2       Rs.   
211942       Akota  22.270027  73.143068  ...           2       Rs.   
211943    Vadiwadi  22.309935  73.158768  ...           2       Rs.   

                                               highlights aggregate_rating  \
211939  ['Dinner', 'Cash', 'Lunch', 'Delivery', 'Indoo...              4.1   
211940  ['Dinner', 'Cash', 'Takeaway Available', 'Debi...              4.1   
211941  ['Dinner', 'Cash', 'Debit Card', 'Lunch', 'Tak...              4.0   
211942  ['Dinner', 'Delivery', 'Credit Card', 'Lunch',...              3.7   
211943  ['Dinner', 'Cash', 'Takeaway Available', 'Debi...              4.0   

       rating_text  votes  photo_count opentable_support delivery  takeaway  
211939   Very Good    243           40               0.0       -1        -1  
211940   Very Good    187           40               0.0        1        -1  
211941   Very Good    111           38               0.0       -1        -1  
211942        Good    128           34               0.0        1        -1  
211943   Very Good     93           53               0.0        1        -1  

[5 rows x 26 columns]

In [9]:
data.head()

res_id                            name      establishment  \
0  3400299                     Bikanervala    ['Quick Bites']   
1  3400005  Mama Chicken Mama Franky House    ['Quick Bites']   
2  3401013                   Bhagat Halwai    ['Quick Bites']   
3  3400290                   Bhagat Halwai    ['Quick Bites']   
4  3401744     The Salt Cafe Kitchen & Bar  ['Casual Dining']   

                                                 url  \
0  https://www.zomato.com/agra/bikanervala-khanda...   
1  https://www.zomato.com/agra/mama-chicken-mama-...   
2  https://www.zomato.com/agra/bhagat-halwai-2-sh...   
3  https://www.zomato.com/agra/bhagat-halwai-civi...   
4  https://www.zomato.com/agra/the-salt-cafe-kitc...   

                                             address  city  city_id  \
0  Kalyani Point, Near Tulsi Cinema, Bypass Road,...  Agra       34   
1        Main Market, Sadar Bazaar, Agra Cantt, Agra  Agra       34   
2  62/1, Near Easy Day, West Shivaji Nagar, Goalp...  Agra       34   
3  Near Anjana Cinema, Nehru Nagar, Civil Lines, ...  Agra       34   
4        1C,3rd Floor, Fatehabad Road, Tajganj, Agra  Agra       34   

      locality   latitude  longitude  ... price_range  currency  \
0     Khandari  27.211450  78.002381  ...           2       Rs.   
1   Agra Cantt  27.160569  78.011583  ...           2       Rs.   
2     Shahganj  27.182938  77.979684  ...           1       Rs.   
3  Civil Lines  27.205668  78.004799  ...           1       Rs.   
4      Tajganj  27.157709  78.052421  ...           3       Rs.   

                                          highlights aggregate_rating  \
0  ['Lunch', 'Takeaway Available', 'Credit Card',...              4.4   
1  ['Delivery', 'No Alcohol Available', 'Dinner',...              4.4   
2  ['No Alcohol Available', 'Dinner', 'Takeaway A...              4.2   
3  ['Takeaway Available', 'Credit Card', 'Lunch',...              4.3   
4  ['Lunch', 'Serves Alcohol', 'Cash', 'Credit Ca...              4.9   

  rating_text  votes  photo_count opentable_support delivery  takeaway  
0   Very Good    814          154               0.0       -1        -1  
1   Very Good   1203          161               0.0       -1        -1  
2   Very Good    801          107               0.0        1        -1  
3   Very Good    693          157               0.0        1        -1  
4   Excellent    470          291               0.0        1        -1  

[5 rows x 26 columns]

In [10]:
data.city.unique()

array(['Agra', 'Ahmedabad', 'Gandhinagar', 'Ajmer', 'Alappuzha',
       'Allahabad', 'Amravati', 'Amritsar', 'Aurangabad', 'Bangalore',
       'Bhopal', 'Bhubaneshwar', 'Chandigarh', 'Mohali', 'Panchkula',
       'Zirakpur', 'Nayagaon', 'Chennai', 'Coimbatore', 'Cuttack',
       'Darjeeling', 'Dehradun', 'New Delhi', 'Gurgaon', 'Noida',
       'Faridabad', 'Ghaziabad', 'Greater Noida', 'Dharamshala',
       'Gangtok', 'Goa', 'Gorakhpur', 'Guntur', 'Guwahati', 'Gwalior',
       'Haridwar', 'Hyderabad', 'Secunderabad', 'Indore', 'Jabalpur',
       'Jaipur', 'Jalandhar', 'Jammu', 'Jamnagar', 'Jamshedpur', 'Jhansi',
       'Jodhpur', 'Junagadh', 'Kanpur', 'Kharagpur', 'Kochi', 'Kolhapur',
       'Kolkata', 'Howrah', 'Kota', 'Lucknow', 'Ludhiana', 'Madurai',
       'Manali', 'Mangalore', 'Manipal', 'Udupi', 'Meerut', 'Mumbai',
       'Thane', 'Navi Mumbai', 'Mussoorie', 'Mysore', 'Nagpur',
       'Nainital', 'Nasik', 'Nashik', 'Neemrana', 'Ooty', 'Palakkad',
       'Patiala', 'Patna', 'Pudu

In [11]:
data.city.nunique()

99

In [12]:
data[data['city'] == "Mumbai"]

res_id                   name      establishment  \
134852  18562340            Drinkery 51  ['Casual Dining']   
134853     34757           Joey's Pizza    ['Quick Bites']   
134854  18548172                Hitchki            ['Bar']   
134855  18392725                Tamasha         ['Lounge']   
134856  18962112               Bayroute    ['Fine Dining']   
...          ...                    ...                ...   
142348     46792       BOX8- Desi Meals                 []   
142349     37750                 Faasos    ['Quick Bites']   
142350  18165909  Charcoal Eats Express    ['Quick Bites']   
142351  18306481            Wok Express     ['Food Court']   
142352  18140577            Rasoi Thali    ['Quick Bites']   

                                                      url  \
134852  https://www.zomato.com/mumbai/drinkery-51-band...   
134853  https://www.zomato.com/mumbai/joeys-pizza-mala...   
134854  https://www.zomato.com/mumbai/hitchki-powai?ut...   
134855  https://www.zomato.com/mumbai/tamasha-lower-pa...   
134856  https://www.zomato.com/mumbai/bayroute-juhu?ut...   
...                                                   ...   
142348  https://www.zomato.com/mumbai/box8-desi-meals-...   
142349  https://www.zomato.com/mumbai/faasos-veera-des...   
142350  https://www.zomato.com/mumbai/charcoal-eats-ex...   
142351  https://www.zomato.com/mumbai/wok-express-1-an...   
142352  https://www.zomato.com/mumbai/rasoi-thali-mulu...   

                                                  address    city  city_id  \
134852  1st Floor, Vibgyor Towers, Bandra Kurla Comple...  Mumbai        3   
134853  Shop 1, Plot D, Samruddhi Complex, Chincholi B...  Mumbai        3   
134854  5, Transocean, Hiranandani Business Park, Lake...  Mumbai        3   
134855  Ground Floor, Victoria House, E.B Marg, Lower ...  Mumbai        3   
134856  14, Silver Beach Estate, A.B. Nair Road, Oppos...  Mumbai        3   
...                                                   ...     ...      ...   
142348  102, Bhavani Industrial Estate, Opposite IIT B...  Mumbai        3   
142349  Shop 06, Building 2, Pramukh Park Rehab CHS Li...  Mumbai        3   
142350         Opposite Kamala Mills, Lower Parel, Mumbai  Mumbai        3   
142351  Food Court, 2nd Floor, Infinity Mall, Unit 215...  Mumbai        3   
142352  Shop 5, Bhodke Building, Near Hanuman Mandir, ...  Mumbai        3   

                                                locality   latitude  \
134852                              Bandra Kurla Complex  19.067176   
134853                                        Malad West  19.178188   
134854                                             Powai  19.119930   
134855                                       Lower Parel  19.006060   
134856                                              Juhu  19.110684   
...                                                  ...        ...   
142348                                             Powai  19.124284   
142349                                  Veera Desai Area  19.139601   
142350                                       Lower Parel  19.002537   
142351  Infiniti Mall, Andheri Lokhandwala, Andheri West  19.141117   
142352                                       Mulund West  19.172657   

        longitude  ... price_range  currency  \
134852  72.867493  ...           4       Rs.   
134853  72.834666  ...           2       Rs.   
134854  72.907331  ...           3       Rs.   
134855  72.827496  ...           4       Rs.   
134856  72.825368  ...           4       Rs.   
...           ...  ...         ...       ...   
142348  72.914877  ...           1       Rs.   
142349  72.833119  ...           1       Rs.   
142350  72.829391  ...           2       Rs.   
142351  72.831181  ...           2       Rs.   
142352  72.956466  ...           1       Rs.   

                                               highlights aggregate_rating  \
134852  ['Dinner', 'Debit Card', 'Lunch', 'Serves Alco...              4.4   
134853  [

In [13]:
data[data['city'] == "Hyderabad"]

res_id                                 name      establishment  \
77501  18392031  Gol Bungalow - Taj Falaknuma Palace    ['Fine Dining']   
77502     91817          Adaa - Taj Falaknuma Palace    ['Fine Dining']   
77503     91818       Celeste - Taj Falaknuma Palace    ['Fine Dining']   
77504  18908475             Vineela Yadlapalli Cakes         ['Bakery']   
77505     92745       Thai Pavilion - Vivanta By Taj    ['Fine Dining']   
...         ...                                  ...                ...   
79616  18402110        The Lal Street - Bar Exchange            ['Bar']   
79617  18649306                             Paradise  ['Casual Dining']   
79618  18459130                 Fat Pigeon - Bar Hop            ['Bar']   
79619     93008        Pride of Hyderabad Restaurant  ['Casual Dining']   
79620  18097695                             Paradise  ['Casual Dining']   

                                                     url  \
77501  https://www.zomato.com/hyderabad/gol-bungalow-...   
77502  https://www.zomato.com/hyderabad/adaa-taj-fala...   
77503  https://www.zomato.com/hyderabad/celeste-taj-f...   
77504  https://www.zomato.com/hyderabad/vineela-yadla...   
77505  https://www.zomato.com/hyderabad/thai-pavilion...   
...                                                  ...   
79616  https://www.zomato.com/hyderabad/the-lal-stree...   
79617  https://www.zomato.com/hyderabad/paradise-2-li...   
79618  https://www.zomato.com/hyderabad/fat-pigeon-ba...   
79619  https://www.zomato.com/hyderabad/pride-of-hyde...   
79620  https://www.zomato.com/hyderabad/paradise-hima...   

                                                 address       city  city_id  \
77501         Taj Falaknuma Palace, Falaknuma, Hyderabad  Hyderabad        6   
77502         Taj Falaknuma Palace, Falaknuma, Hyderabad  Hyderabad        6   
77503         Taj Falaknuma Palace, Falaknuma, Hyderabad  Hyderabad        6   
77504  Plot 805, Behind B-DUBS, Road Number 36, CBI C...  Hyderabad        6   
77505  Vivanta By Taj, 1-10-147 & 148, Mayuri Marg, B...  Hyderabad        6   
...                                                  ...        ...      ...   
79616  Level 2, SLN Terminus Mall, Gachibowli, Hyderabad  Hyderabad        6   
79617  23-19 & 23-20, Survey 204/2, Near BHEL, Near J...  Hyderabad        6   
79618     Plot 1335/A, Road 45, Jubilee Hills, Hyderabad  Hyderabad        6   
79619                  Bharat Nagar, Moosapet, Hyderabad  Hyderabad        6   
79620  Ground Floor, Lingapur Building, Next to Almon...  Hyderabad        6   

                       locality   latitude  longitude  ... price_range  \
77501  Taj Falaknuma, Falaknuma  17.334338  78.467603  ...           4   
77502  Taj Falaknuma, Falaknuma  17.334283  78.467568  ...           4   
77503  Taj Falaknuma, Falaknuma  17.334312  78.467562  ...           4   
77504             Jubilee Hills  17.433280  78.404894  ...           4   
77505  Vivanta By Taj, Begumpet  17.443687  78.460743  ...           4   
...                         ...        ...        ...  ...         ...   
79616   SLN Terminus,Gachibowli  17.452475  78.363198  ...           3   
79617               Lingampally  17.502396  78.312111  ...           2   
79618             Jubilee Hills  17.425610  78.408536  ...           3   
79619                  Moosapet  17.467529  78.429099  ...           2   
79620            Himayath Nagar  17.401393  78.485650  ...           2   

       currency                                         highlights  \
77501       Rs.  ['Cash', 'Dinner', 'Credit Card', 'No Alcohol ...   
77502       Rs.  ['Lunch', 'Cash', 'Credit Card', 'Dinner', 'De...   
77503       Rs.  ['Breakfast', 'Lunch', 'Cash', 'Debit Card', '...   
77504       Rs.  ['Cash', 'Takeaway Available', 'Debit Card', '...   
77505       Rs.  ['Lunch', 'Cash', 'Debit Card', 'Dinner', 'Cre...   
...         ...                                                ...   
79616       Rs.  ['Lunch', 'Debit Card', 'Cash'

In [14]:
data.shape

(211944, 26)

our data has 2 million plus rows and 26 columns.

Let us go and understand what are all there in these rows and columns

In [15]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 211944 entries, 0 to 211943
Data columns (total 26 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   res_id                211944 non-null  int64  
 1   name                  211944 non-null  object 
 2   establishment         211944 non-null  object 
 3   url                   211944 non-null  object 
 4   address               211810 non-null  object 
 5   city                  211944 non-null  object 
 6   city_id               211944 non-null  int64  
 7   locality              211944 non-null  object 
 8   latitude              211944 non-null  float64
 9   longitude             211944 non-null  float64
 10  zipcode               48757 non-null   object 
 11  country_id            211944 non-null  int64  
 12  locality_verbose      211944 non-null  object 
 13  cuisines              210553 non-null  object 
 14  timings               208070 non-null  object 
 15  

Here, we are dealing with the large amount of the data, so it is better to find out what is the count, null and type properties of the columns with the help of info()

In [16]:
##statistical analysis can be done with the help of describe()

data.describe()

res_id        city_id       latitude      longitude  country_id  \
count  2.119440e+05  211944.000000  211944.000000  211944.000000    211944.0   
mean   1.349411e+07    4746.785434      21.499758      77.615276         1.0   
std    7.883722e+06    5568.766386      22.781331       7.500104         0.0   
min    5.000000e+01       1.000000       0.000000       0.000000         1.0   
25%    3.301027e+06      11.000000      15.496071      74.877961         1.0   
50%    1.869573e+07      34.000000      22.514494      77.425971         1.0   
75%    1.881297e+07   11306.000000      26.841667      80.219323         1.0   
max    1.915979e+07   11354.000000   10000.000000      91.832769         1.0   

       average_cost_for_two    price_range  aggregate_rating          votes  \
count         211944.000000  211944.000000     211944.000000  211944.000000   
mean             595.812229       1.882535          3.395937     378.001864   
std              606.239363       0.892989          1.283642     925.333370   
min                0.000000       1.000000          0.000000     -18.000000   
25%              250.000000       1.000000          3.300000      16.000000   
50%              400.000000       2.000000          3.800000     100.000000   
75%              700.000000       2.000000          4.100000     362.000000   
max            30000.000000       4.000000          4.900000   42539.000000   

         photo_count  opentable_support       delivery  takeaway  
count  211944.000000           211896.0  211944.000000  211944.0  
mean      256.971224                0.0      -0.255907      -1.0  
std       867.668940                0.0       0.964172       0.0  
min         0.000000                0.0      -1.000000      -1.0  
25%         3.000000                0.0      -1.000000      -1.0  
50%        18.000000                0.0      -1.000000      -1.0  
75%       128.000000                0.0       1.000000      -1.0  
max     17702.000000                0.0       1.000000      -1.0

As the dataset that we have taken may be having duplicate values, try to understand this : when we are filling out the form we may fill one or more times and ending up increase the amount of responses.  So it is better to remove all the duplicates before getting in to the analysis part.

In [17]:
data.drop_duplicates(['res_id'],keep = 'first', inplace = True)

In [18]:
data.shape

(55568, 26)

It is crazy that the dataset that we have taken has so many dupliactes, it is good that we had removed duplicates before performing the analysis. And now we are left with one fourh of the data. It is good enough to perform analysis on 50k+ rows.

# Dealing with the missing values.



In [19]:
data.isnull()

res_id   name  establishment    url  address   city  city_id  \
0        False  False          False  False    False  False    False   
1        False  False          False  False    False  False    False   
2        False  False          False  False    False  False    False   
3        False  False          False  False    False  False    False   
4        False  False          False  False    False  False    False   
...        ...    ...            ...    ...      ...    ...      ...   
211882   False  False          False  False    False  False    False   
211925   False  False          False  False    False  False    False   
211926   False  False          False  False    False  False    False   
211940   False  False          False  False    False  False    False   
211942   False  False          False  False    False  False    False   

        locality  latitude  longitude  ...  price_range  currency  highlights  \
0          False     False      False  ...        False     False       False   
1          False     False      False  ...        False     False       False   
2          False     False      False  ...        False     False       False   
3          False     False      False  ...        False     False       False   
4          False     False      False  ...        False     False       False   
...          ...       ...        ...  ...          ...       ...         ...   
211882     False     False      False  ...        False     False       False   
211925     False     False      False  ...        False     False       False   
211926     False     False      False  ...        False     False       False   
211940     False     False      False  ...        False     False       False   
211942     False     False      False  ...        False     False       False   

        aggregate_rating  rating_text  votes  photo_count  opentable_support  \
0                  False        False  False        False              False   
1                  False        False  False        False              False   
2                  False        False  False        False              False   
3                  False        False  False        False              False   
4                  False        False  False        False              False   
...                  ...          ...    ...          ...                ...   
211882             False        False  False        False              False   
211925             False        False  False        False              False   
211926             False        False  False        False              False   
211940             False        False  False        False              False   
211942             False        False  False        False              False   

        delivery  takeaway  
0          False     False  
1          False     False  
2          False     False  
3          False     False  
4          False     False  
...          ...       ...  
211882     False     False  
211925     False     False  
211926     False     False  
211940     False     False  
211942     False     False  

[55568 rows x 26 columns]

In [20]:
data.isnull().sum()

res_id                      0
name                        0
establishment               0
url                         0
address                    18
city                        0
city_id                     0
locality                    0
latitude                    0
longitude                   0
zipcode                 44623
country_id                  0
locality_verbose            0
cuisines                  470
timings                  1003
average_cost_for_two        0
price_range                 0
currency                    0
highlights                  0
aggregate_rating            0
rating_text                 0
votes                       0
photo_count                 0
opentable_support          12
delivery                    0
takeaway                    0
dtype: int64

We are having 5 features that has missingvalues, since the zipcode has more number of missing values(75%), so we can simply ignore that featurein our analysis.

**Omitting or not considering the useful features**

Now, we need to look at each and every feature and has to decide whether they are useful for our analysis or not.


1. res_id - Unique id for each restuarant, so need to worry about this column

2. name - Name is useful in our analysis to find the best and top restaurants

3. establishment - let us see what this feature actually represents

In [21]:
data['establishment']

0             ['Quick Bites']
1             ['Quick Bites']
2             ['Quick Bites']
3             ['Quick Bites']
4           ['Casual Dining']
                 ...         
211882    ['Dessert Parlour']
211925      ['Casual Dining']
211926      ['Casual Dining']
211940        ['Quick Bites']
211942        ['Quick Bites']
Name: establishment, Length: 55568, dtype: object

In [22]:
#checking out the unique values

data['establishment'].unique()

array(["['Quick Bites']", "['Casual Dining']", "['Bakery']", "['Café']",
       "['Dhaba']", "['Bhojanalya']", "['Bar']", "['Sweet Shop']",
       "['Fine Dining']", "['Food Truck']", "['Dessert Parlour']",
       "['Lounge']", "['Pub']", "['Beverage Shop']", "['Kiosk']",
       "['Paan Shop']", "['Confectionery']", '[]', "['Shack']",
       "['Club']", "['Food Court']", "['Mess']", "['Butcher Shop']",
       "['Microbrewery']", "['Cocktail Bar']", "['Pop up']",
       "['Irani Cafe']"], dtype=object)

In [23]:
data['establishment'].nunique()

27

In [24]:
print(data['establishment'].unique()[0])

['Quick Bites']


In [25]:
type(data['establishment'].unique()[0])

str

Here, establishment looks like an interesting feature to perform the analysis, it is sad that we are finding out the unnecessary brackets for the values in feature establishment.

Our main task is to remove those unwanted brackets ([''])from the values

In [26]:
#Removing [' '] from the each value

data['establishment'] = data['establishment'].apply(lambda x:x[2:-2])
print(data['establishment'].unique())

['Quick Bites' 'Casual Dining' 'Bakery' 'Café' 'Dhaba' 'Bhojanalya' 'Bar'
 'Sweet Shop' 'Fine Dining' 'Food Truck' 'Dessert Parlour' 'Lounge' 'Pub'
 'Beverage Shop' 'Kiosk' 'Paan Shop' 'Confectionery' '' 'Shack' 'Club'
 'Food Court' 'Mess' 'Butcher Shop' 'Microbrewery' 'Cocktail Bar' 'Pop up'
 'Irani Cafe']


Yeahh we had removed the unnecessary brackets from the values, now its our turn to replace the " " empty values with NA

In [27]:
#Changing ' ' to NA

data['establishment'] = data['establishment'].apply(lambda x: np.where(x == "","NA",x))
print(data['establishment'].unique())

['Quick Bites' 'Casual Dining' 'Bakery' 'Café' 'Dhaba' 'Bhojanalya' 'Bar'
 'Sweet Shop' 'Fine Dining' 'Food Truck' 'Dessert Parlour' 'Lounge' 'Pub'
 'Beverage Shop' 'Kiosk' 'Paan Shop' 'Confectionery' 'NA' 'Shack' 'Club'
 'Food Court' 'Mess' 'Butcher Shop' 'Microbrewery' 'Cocktail Bar' 'Pop up'
 'Irani Cafe']


### We are done with the establishment feature
Now, let's move forward for another features

4. url - URL is the link for the restaurant which is not useful in our analysis

5. address - not useful for us and it is containing long strings, which is difficult for use and analyse.

6. city - let's check what are the unique vales in city

In [28]:
data['city'].unique()

array(['Agra', 'Ahmedabad', 'Gandhinagar', 'Ajmer', 'Alappuzha',
       'Allahabad', 'Amravati', 'Amritsar', 'Aurangabad', 'Bangalore',
       'Bhopal', 'Bhubaneshwar', 'Chandigarh', 'Mohali', 'Panchkula',
       'Zirakpur', 'Nayagaon', 'Chennai', 'Coimbatore', 'Cuttack',
       'Darjeeling', 'Dehradun', 'New Delhi', 'Gurgaon', 'Noida',
       'Faridabad', 'Ghaziabad', 'Greater Noida', 'Dharamshala',
       'Gangtok', 'Goa', 'Gorakhpur', 'Guntur', 'Guwahati', 'Gwalior',
       'Haridwar', 'Hyderabad', 'Secunderabad', 'Indore', 'Jabalpur',
       'Jaipur', 'Jalandhar', 'Jammu', 'Jamnagar', 'Jamshedpur', 'Jhansi',
       'Jodhpur', 'Junagadh', 'Kanpur', 'Kharagpur', 'Kochi', 'Kolhapur',
       'Kolkata', 'Howrah', 'Kota', 'Lucknow', 'Ludhiana', 'Madurai',
       'Manali', 'Mangalore', 'Manipal', 'Udupi', 'Meerut', 'Mumbai',
       'Thane', 'Navi Mumbai', 'Mussoorie', 'Mysore', 'Nagpur',
       'Nainital', 'Nasik', 'Nashik', 'Neemrana', 'Ooty', 'Palakkad',
       'Patiala', 'Patna', 'Pudu

In [29]:
len(data['city'].unique())

99

In [30]:
data[data['city'] ==  "Goa"]

res_id                       name  establishment  \
63394  16512333  Britto's Bar & Restaurant  Casual Dining   
63395    130274                   Thalassa  Casual Dining   
63396    130535      The Fisherman's Wharf  Casual Dining   
63397    130309             Purple Martini  Casual Dining   
63398    130409               Baba Au Rhum  Casual Dining   
...         ...                        ...            ...   
66370  18577432     304 Kitchen And Tavern  Casual Dining   
66371    130442              Avalon Sunset  Casual Dining   
66372  18805479                   Flamingo  Casual Dining   
66376  18964410             Chilli n Spice    Quick Bites   
66377  18233522                     Darbar  Casual Dining   

                                                     url  \
63394  https://www.zomato.com/goa/brittos-bar-restaur...   
63395  https://www.zomato.com/goa/thalassa-siolim?utm...   
63396  https://www.zomato.com/goa/the-fishermans-whar...   
63397  https://www.zomato.com/goa/purple-martini-anju...   
63398  https://www.zomato.com/goa/baba-au-rhum-1-anju...   
...                                                  ...   
66370  https://www.zomato.com/goa/304-kitchen-and-tav...   
66371  https://www.zomato.com/goa/avalon-sunset-anjun...   
66372  https://www.zomato.com/goa/flamingo-panaji?utm...   
66376  https://www.zomato.com/goa/chilli-n-spice-anju...   
66377  https://www.zomato.com/goa/darbar-panaji?utm_s...   

                                                 address city  city_id  \
63394                  Baga Calangute, Bardez, Baga, Goa  Goa       13   
63395  Near Teso Waterfront, Plot 301, 1, Vaddy, Siol...  Goa       13   
63396           Before The Leela, Mobor, Cavelossim, Goa  Goa       13   
63397                          Anjuna Beach, Anjuna, Goa  Goa       13   
63398                       1054, Sim Vaddo, Anjuna, Goa  Goa       13   
...                                                  ...  ...      ...   
66370  Opposite Casa Menezes - The Goan Heritage Home...  Goa       13   
66371                          Anjuna Beach, Anjuna, Goa  Goa       13   
66372                               Miramar, Panaji, Goa  Goa       13   
66376             H 211/D 3, Chivar, Bardez, Anjuna, Goa  Goa       13   
66377  Hotel Goa International, D.B Marg,Near Miramar...  Goa       13   

                   locality   latitude  longitude  ... price_range  currency  \
63394                  Baga  15.560983  73.749268  ...           4       Rs.   
63395                Siolim  15.616222  73.755632  ...           4       Rs.   
63396            Cavelossim  15.157532  73.952245  ...           4       Rs.   
63397  Anjuna Beach, Anjuna  15.583275  73.737320  ...           4       Rs.   
63398                Anjuna  15.576377  73.755765  ...           4       Rs.   
...                     ...        ...        ...  ...         ...       ...   
66370              Bambolim  15.449189  73.886606  ...           4       Rs.   
66371  Anjuna Beach, Anjuna  15.581365  73.737981  ...           3       Rs.   
66372                Panaji  15.464708  73.807543  ...           3       Rs.   
66376                Anjuna  15.600046  73.751396  ...           1       Rs.   
66377                Panaji  15.482725  73.811701  ...           4       Rs.   

                                              highlights aggregate_rating  \
63394  ['Takeaway Available', 'Credit Card', 'Serves ...              4.4   
63395  ['Debit Card', 'Dinner', 'Credit Card', 'Lunch...              4.4   
63396  ['Dinner', 'Debit Card', 'Serves Alcohol', 'Cr...              4.8   
63397  ['Dinner', 'Credit Card', 'Lunch', 'Cash', 'Ta...              4.9   
63398  ['Dinner', 'Cash', 'Credit Card', 'Debit Card'...              4.7   
...                                                  ...              ...   
66370  ['Serves Alcohol', 'Breakfast', 'Dinner', 'Cas...              3.7   
66371  ['Lunch', 'Cash', 'Breakfast', 'Dinner', 'Full...              3.6   
66372  ['Cre

7. city_id - We can either use city id or cityname so one feature is enough

8. locality - Locality and city sounds like same, but lets us see what is there in locality feature

In [31]:
data['locality'].head()

0       Khandari
1     Agra Cantt
2       Shahganj
3    Civil Lines
4        Tajganj
Name: locality, dtype: object

In [32]:
#checking out the unique values

data['locality'].unique()

array(['Khandari', 'Agra Cantt', 'Shahganj', ..., 'Navapura',
       'L&T Knowledge City', 'Danteshwar'], dtype=object)

In [33]:
data['locality'].nunique()

3731

We are having so many unique values, we can avoid it besides it is an interesting feature for our analysis.

9. latitude - useful for geographical analysis, we won't be doing that here.

10. longitude - same as above

11. zipcode - approx 75% of the missing values are present

12. country_id - As this dataset that we have taken is only for the indian restaurants. There should be only one country code. Let us check

In [34]:
data['country_id'].unique()

array([1])

13. locality_verbose - let's check this feature

In [35]:
data['locality_verbose'].head()

0       Khandari, Agra
1     Agra Cantt, Agra
2       Shahganj, Agra
3    Civil Lines, Agra
4        Tajganj, Agra
Name: locality_verbose, dtype: object

In [36]:
data['locality_verbose'].tail()

211882     Manjalpur, Vadodara
211925      Alkapuri, Vadodara
211926      Alkapuri, Vadodara
211940    Karelibaug, Vadodara
211942         Akota, Vadodara
Name: locality_verbose, dtype: object

### Here locality and locality verbose looks same

In [37]:
data['locality_verbose'].nunique()

3910

In [38]:
data['cuisines']

0         North Indian, South Indian, Mithai, Street Foo...
1         North Indian, Mughlai, Rolls, Chinese, Fast Fo...
2                                         Fast Food, Mithai
3                 Desserts, Bakery, Fast Food, South Indian
4                        North Indian, Continental, Italian
                                ...                        
211882                                            Ice Cream
211925                      Gujarati, North Indian, Chinese
211926                                Gujarati, Street Food
211940                                            Fast Food
211942                           Fast Food, Sandwich, Salad
Name: cuisines, Length: 55568, dtype: object

In [39]:
data['cuisines'].unique()

array(['North Indian, South Indian, Mithai, Street Food, Desserts',
       'North Indian, Mughlai, Rolls, Chinese, Fast Food, Street Food',
       'Fast Food, Mithai', ...,
       'Street Food, Biryani, Chinese, Fast Food, North Indian, Mughlai',
       'North Indian, Chinese, Mexican, Italian, Thai, Continental',
       'North Indian, Lucknowi, Chinese'], dtype=object)

In [40]:
data['cuisines'].nunique()

9382

14. cuisines - This feature is having some missing values. But we are having 9382 unique values. the reason for this is that each restaurant has a list of cuisines and the addition of all those is resulting in the huge number of unique values.

a) fill the null values

b) check number of unique cuisines

In [41]:
data['cuisines'] = data['cuisines'].fillna("No cuisine")

In [42]:
data['cuisines'].isnull().sum()

np.int64(0)

In [43]:
cuisines = []
data['cuisines'].apply(lambda x:cuisines.extend(x.split(",")))
cuisines = pd.Series(cuisines)
print("Total unique cuisines =  ", cuisines.nunique())

Total unique cuisines =   249


In [44]:
data['cuisines']

0         North Indian, South Indian, Mithai, Street Foo...
1         North Indian, Mughlai, Rolls, Chinese, Fast Fo...
2                                         Fast Food, Mithai
3                 Desserts, Bakery, Fast Food, South Indian
4                        North Indian, Continental, Italian
                                ...                        
211882                                            Ice Cream
211925                      Gujarati, North Indian, Chinese
211926                                Gujarati, Street Food
211940                                            Fast Food
211942                           Fast Food, Sandwich, Salad
Name: cuisines, Length: 55568, dtype: object

In [45]:
data['timings']

0                                8:30am – 10:30pm (Mon-Sun)
1         12:30PM to 12Midnight (Mon, Wed, Thu, Fri, Sat...
2                                          9:30 AM to 11 PM
3                                      8am – 11pm (Mon-Sun)
4                                      11:30 AM to 11:30 PM
                                ...                        
211882                                 11 AM to 12 Midnight
211925                11:30 AM to 3:30 PM, 7:30 PM to 11 PM
211926                   11 AM to 3:30 PM, 7 PM to 10:30 PM
211940    4pm – 11pm (Mon, Tue, Wed, Fri, Sat, Sun), 4pm...
211942                                         8 AM to 1 AM
Name: timings, Length: 55568, dtype: object

In [46]:
data['timings'].unique()

array(['8:30am – 10:30pm (Mon-Sun)',
       '12:30PM to 12Midnight (Mon, Wed, Thu, Fri, Sat, Sun), 1PM to 12Midnight (Tue)',
       '9:30 AM to 11 PM', ..., '8am – 2:30pm, 5pm – 10pm (Mon-Sun)',
       '8am – 3pm, 6:30pm – 11pm (Mon-Sun)',
       '4pm – 11pm (Mon, Tue, Wed, Fri, Sat, Sun), 4pm – 11:30pm (Thu)'],
      dtype=object)

In [47]:
data['timings'].nunique()

7740

15. Timings - This feature is also having the missing values, but we are having large number of unique classes. and it is not even structured properly. better we avoid this feature

16. average_cost_of_two - this is an excellent feature, we are having value "0", this should be an outlier.

In [48]:
data['average_cost_for_two'].nunique()

145

17. price_range - let us know more about this.

In [49]:
data['price_range']

0         2
1         2
2         1
3         1
4         3
         ..
211882    1
211925    2
211926    1
211940    1
211942    2
Name: price_range, Length: 55568, dtype: int64

In [50]:
data['price_range'].unique()

array([2, 1, 3, 4])

In [51]:
data['price_range'].nunique()

4

18. currency - As we are dealing with the indian restuarant, currency should be only one unique value

In [52]:
data['currency']

0         Rs.
1         Rs.
2         Rs.
3         Rs.
4         Rs.
         ... 
211882    Rs.
211925    Rs.
211926    Rs.
211940    Rs.
211942    Rs.
Name: currency, Length: 55568, dtype: object

In [53]:
data['currency'].unique()

array(['Rs.'], dtype=object)

In [54]:
len(data['currency'].unique())

1

In [55]:
data['highlights']

0         ['Lunch', 'Takeaway Available', 'Credit Card',...
1         ['Delivery', 'No Alcohol Available', 'Dinner',...
2         ['No Alcohol Available', 'Dinner', 'Takeaway A...
3         ['Takeaway Available', 'Credit Card', 'Lunch',...
4         ['Lunch', 'Serves Alcohol', 'Cash', 'Credit Ca...
                                ...                        
211882    ['Cash', 'Takeaway Available', 'Delivery', 'In...
211925    ['Dinner', 'Cash', 'Debit Card', 'Lunch', 'Tak...
211926    ['Dinner', 'Cash', 'Credit Card', 'Lunch', 'Ta...
211940    ['Dinner', 'Cash', 'Takeaway Available', 'Debi...
211942    ['Dinner', 'Delivery', 'Credit Card', 'Lunch',...
Name: highlights, Length: 55568, dtype: object

19. Highlights - This feature represents the specialities of the restaurants and want to make their attention more in the market. Same like cuisines feature here also we need to follow the same steps and decompose the singe feature values, to find the unique values of this feature


In [56]:
data['highlights'].unique()

array(["['Lunch', 'Takeaway Available', 'Credit Card', 'Dinner', 'Cash', 'Air Conditioned', 'Indoor Seating', 'Pure Veg']",
       "['Delivery', 'No Alcohol Available', 'Dinner', 'Takeaway Available', 'Lunch', 'Cash', 'Indoor Seating']",
       "['No Alcohol Available', 'Dinner', 'Takeaway Available', 'Breakfast', 'Lunch', 'Cash', 'Delivery', 'Outdoor Seating', 'Air Conditioned', 'Self Service', 'Indoor Seating', 'Digital Payments Accepted', 'Pure Veg', 'Desserts and Bakes']",
       ...,
       "['Dinner', 'Delivery', 'Cash', 'Takeaway Available', 'Free Parking', 'Digital Payments Accepted', 'Pure Veg', 'Indoor Seating']",
       "['Dinner', 'Cash', 'Takeaway Available', 'Lunch', 'Delivery', 'Free Parking', 'Indoor Seating', 'Air Conditioned', 'Outdoor Seating', 'Digital Payments Accepted', 'Catering Available', 'Pure Veg']",
       "['Dinner', 'Cash', 'Takeaway Available', 'Debit Card', 'Delivery', 'Credit Card', 'Free Parking', 'Outdoor Seating']"],
      dtype=object)

In [59]:
data['highlights'].nunique()

31228

In [60]:
h1=[ ]
data['highlights'].apply(lambda x:h1.extend(x.split(",")))
h1=pd.Series(h1)
print("Total unique highlights =  ", h1.nunique())

Total unique highlights =   244


20. aggregate_rating - Rating that restaurant has

21. rating_text = This can be taken in the form of bins using labels, we can use direct aggregate_rating in our analysis. So we can ignore this feature

22. votes - Number of votes contribute the rating

23. Photo_count - Number of photos uploaded at the time of reviews.


All the above features (20,22,23) have int values, so we can perform some analysis for these features

In [61]:
data[['aggregate_rating', 'votes', 'photo_count']].describe()

aggregate_rating         votes   photo_count
count      55568.000000  55568.000000  55568.000000
mean           2.958593    223.330352    160.974770
std            1.464576    618.224019    586.975382
min            0.000000    -18.000000      0.000000
25%            2.900000      6.000000      1.000000
50%            3.500000     35.000000     10.000000
75%            3.900000    175.000000     69.000000
max            4.900000  42539.000000  17702.000000

In [62]:
data[['aggregate_rating', 'votes', 'photo_count']].describe().loc[['mean','max','min']]

aggregate_rating         votes  photo_count
mean          2.958593    223.330352    160.97477
max           4.900000  42539.000000  17702.00000
min           0.000000    -18.000000      0.00000

## From the above analysis we can say that:

Zomato ratings are between 0 to 5

Maximum number of votes for a restaurant in Zomato - 42539votes, while we are getting -18 minimum votes this might be an outlier.

And there is one particular restaurant which is having maximum votes and maximum photos at the time of review .

In [64]:
data['opentable_support']

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
         ... 
211882    0.0
211925    0.0
211926    0.0
211940    0.0
211942    0.0
Name: opentable_support, Length: 55568, dtype: float64

In [66]:
data['delivery']

0        -1
1        -1
2         1
3         1
4         1
         ..
211882    1
211925   -1
211926   -1
211940    1
211942    1
Name: delivery, Length: 55568, dtype: int64

In [67]:
data['delivery'].unique()

array([-1,  1,  0])

In [68]:
data['takeaway']

0        -1
1        -1
2        -1
3        -1
4        -1
         ..
211882   -1
211925   -1
211926   -1
211940   -1
211942   -1
Name: takeaway, Length: 55568, dtype: int64

In [70]:
data['takeaway'].unique()

array([-1])

24. opentable_support - we are not having any true value for this, better we ignore this.

25. Delivery - This feature is having three classes, but there is no proper explanation for this classes, ignore this feature for now.

26. takeaway - not useful as it is having only one class.